src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -1218,71 +1250,131 @@ void ObjectMonitor::VThreadEpilog(JavaThread* current, ObjectWaiter* node) {
   }
 }
 
-// By convention we unlink a contending thread from EntryList|cxq immediately
-// after the thread acquires the lock in ::enter().  Equally, we could defer
-// unlinking the thread until ::exit()-time.
+// Return the tail of the _entry_list. If the tail is currently not
+// known, find it by walking from the head of _entry_list, and while
+// doing so assign the _prev pointers to create a doubly linked list.
+ObjectWaiter* ObjectMonitor::entry_list_tail(JavaThread* current) {
+  assert(has_owner(current), "invariant");
+  ObjectWaiter* w = _entry_list_tail;
+  if (w != nullptr) {
+    return w;
+  }
+  // Need acquire here to match the implicit release of the cmpxchg
+  // that updated _entry_list, so we can access w->_next.
+  w = Atomic::load_acquire(&_entry_list);
+  assert(w != nullptr, "invariant");
```

Comment:
Maybe add the same assert as below for the single element case: `assert(w->TState == ObjectWaiter::TS_ENTER, "invariant")`.